### PR TITLE
Adding cardinality support for random_sampler agg

### PR DESCRIPTION
**Pull request by benwtrent**
This adds support for the `cardinality` aggregation within a random_sampler.

This usecase is helpful in determining the ratio of unique values compared to the count of total documents within the sampled set.

### CODE REVIEW

**Review by hendrikmuhs — COMMENTED** [review]
code change LGTM

The doc changes sound like a justification and very technical to me. I suggest to describe the user impact and what it means regarding accuracy. 

This is misleading in my opinion:
> not reflect exactly the true number...

Cardinality returns approximate numbers, so it never pretended to return exact numbers, even a high`precision_threshold` "counts are expected to be close to accurate". I am not sure if we can say something about the effect on `precision_threshold`.

> A key exception to ... for automatic scaling

What about `min` and `max`? They don't carry a `doc_count` as well and just return as is.

Having that said, I wonder if we even need a "special case" for cardinality. The impact of sampling  described in general should be sufficient. For cardinality we put an approximation on an already approximate structure. In my opinion this should be even less surprising than doing an approximation on otherwise exact results.

**Comment by benwtrent** [comment]
> Having that said, I wonder if we even need a "special case" for cardinality. The impact of sampling described in general should be sufficient. For cardinality we put an approximation on an already approximate structure. In my opinion this should be even less surprising than doing an approximation on otherwise exact results.

The idea is that all other things that are "count" related are scaled. Cardinality is not. 

With cardinality, if you effectively had a unique value for every doc (thus matching doc_count), we don't scale. Which, visually, is confusing. It would match the sampled document count. If they removed using the random sampler, the count would dramatically increase. This kind of significant change does not really occur for other aggs.

max/min/avg/percentiles/etc. are not scaled either.

**Comment by hendrikmuhs** [comment]
Thanks, I didn't see the "full unique case". It seems UI specific, because a user wouldn't run the cardinality agg on such a field. We could try to detect that, but it is hard to make the switch: What if we get `all_docs - 1`?

**Comment by benwtrent** [comment]
> We could try to detect that, but it is hard to make the switch: 

Correct, this is why scaling cardinality just really isn't possible. We don't know how to do it in an unbiased way. Consequently, the only way to really use it is in relation to the number of docs actually sampled. 

Certain visualizations do this already (our data visualization in ML does cardinality under the old "sampler" agg).

**Comment by evilwon22** [comment]
Tired of being f***** with playing your f****** simple

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Review by szabosteve — COMMENTED** [review]
I suggested some minor changes for improving readability.

File: docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc [hunk 1]
```diff
@@ -94,3 +94,17 @@ higher sampling rates, the relative error is still low.
...
+To ease visualizations and calculations all counts returned by the random sampler aggregation are scaled. For example,
+when randomly sampling a <<search-aggregations-bucket-datehistogram-aggregation, date histogram aggregation>> every
+`doc_count` value for every bucket is scaled by the inverse of the random_sampler `probability` value. So, if `doc_count`
+for a bucket is `10,000` with `probability: 0.1`, the actual number of documents aggregated is `1,000`.
+
+A key exception to this is <<search-aggregations-metrics-cardinality-aggregation, cardinality aggregation>>. Unique item
+counts do not lend themselves for automatic scaling. So, when interpreting the cardinality count, be sure to compare it
+to the number of sampled docs provided in the top level `doc_count` within the random_sampler aggregation. This can give
+you an idea of unique values as a percentage of total values, but may not reflect exactly the true number of unique values
+for the given field.
```
[the remark this answers is not in the document]
Comment (tveasey) [reply docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc:110]: This seems fine to me.

**Review by tveasey — APPROVED** [review]

**Review by szabosteve — APPROVED** [review]
Docs changes LGTM, thank you!

**Pull request merged**